### PR TITLE
feat: add createLocalStorage to enable custom storage namespace

### PR DESCRIPTION
**Pull request by heyacherry**
Feat:
--
- Extend the flexibility to add custom storageNamespace, 
- Related discussion here:  
   [0] https://github.com/graphql/graphiql/issues/2794  
   [1] https://github.com/graphql/graphiql/pull/3021

### CODE REVIEW

**Comment by CaiYiLiang** [comment]
Hello @acao @thomasheyenbrock 
Not sure whether this change is what you prefer? Pls let me know if any other ideas!

File: packages/graphiql-toolkit/src/storage/custom.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/**
+ * This function enable a custom namespace for localStorage
```
Comment (TallTed) [inline packages/graphiql-toolkit/src/storage/custom.ts:2]:
```suggestion
 * This function enables a custom namespace for localStorage
```

**Review by TallTed — CHANGES_REQUESTED** [review]

File: packages/graphiql-toolkit/src/storage/custom.ts (hunk 1 above)
Comment (CaiYiLiang) [reply packages/graphiql-toolkit/src/storage/custom.ts]: Thank you! I have updated! @TallTed

**Comment by acao** [comment]
@CaiYiLiang thanks for this! not sure what is up with the CLA bot - it always comes down to the email address you used to make the commits with your git client. do you want to close #3021 now that this is passing?

**Comment by CaiYiLiang** [comment]
> @CaiYiLiang thanks for this! not sure what is up with the CLA bot - it always comes down to the email address you used to make the commits with your git client. do you want to close #3021 now that this is passing?

Hello @acao I have closed #3021 .
Please let me know if there is anything I can help!

File: packages/graphiql-toolkit/src/storage/custom.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/**
+ * This function enables a custom namespace for localStorage
+ */
+
+import { Storage } from './base';
+
+export function createLocalStorage(namespace: string): Storage {
+  const storage = {
```
Comment (acao) [inline packages/graphiql-toolkit/src/storage/custom.ts:8]:
```suggestion
  // you can re-use this key in the methods below
  const storageKey = `${namespace}:${key}`;
  const storage = {
```

Comment (CaiYiLiang) [reply packages/graphiql-toolkit/src/storage/custom.ts:8]: Yes, that's the part definely shoule be improved! Thanks @acao

File: .changeset/lovely-glasses-cross.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@graphiql/toolkit': patch
```
Comment (thomasheyenbrock) [inline .changeset/lovely-glasses-cross.md:2]: Let's make this a minor as we're adding new functionality

File: .changeset/lovely-glasses-cross.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+---
+
+add createLocalStorage util for custom storage namespace
```
Comment (thomasheyenbrock) [inline .changeset/lovely-glasses-cross.md:5]:
```suggestion
Add a new utility function `createLocalStorage` that creates a local storage with support for custom namespaces
```

File: packages/graphiql-toolkit/src/storage/custom.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/**
+ * This function enables a custom namespace for localStorage
+ */
+
+import { Storage } from './base';
+
+export function createLocalStorage(namespace: string): Storage {
+  // you can re-use storageKeyPrefix/storageKey in the methods below
```
Comment (thomasheyenbrock) [inline packages/graphiql-toolkit/src/storage/custom.ts:8]: I don't think we need that comment, variables are always meant to be reused.
```suggestion
```

**Review by thomasheyenbrock — CHANGES_REQUESTED** [review]
Looking good 👏  just some minor nits

File: packages/graphiql-toolkit/src/storage/custom.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
+/**
+ * This function enables a custom namespace for localStorage
+ */
+
+import { Storage } from './base';
+
+export function createLocalStorage(namespace: string): Storage {
```
Comment (acao) [inline packages/graphiql-toolkit/src/storage/custom.ts:7]: I still would prefer this to be an object, so it's easier to expand on, and we don't have to worry about ordered arguments as further options are added.
```suggestion
export CreateLocalStorageOptions = {
  /**
   * specify a different storage namespace prefix from the default of 'graphiql' 
   */
  namespace?: string
}
/**
 * generate a custom local storage adapter for GraphiQL `storage` prop.
 */
export function createLocalStorage({ namespace }: ): Storage {
```

either way, this library should have some API docs added like this!

**Comment by acao** [comment]
@heyacherry do you mind if I make a few last commits to make this ready for 3.0.0? or would prefer to finish the PR?

File: .changeset/lovely-glasses-cross.md (hunk 3 above)
Comment (acao) [reply .changeset/lovely-glasses-cross.md:2]:
```suggestion
'@graphiql/toolkit': minor
```

**Comment by acao** [comment]
I don't know whats wrong, propbably something on my end, but it won't let me push a local commit and rebase to your branch

image

**Comment by acao** [comment]
@B2o5T looks like we have test flakiness with the `monaco-graphql-react-vite` example again. this test was passing on the last release, but now on rebase with a fresh `yarn`, we are getting an additional module - I think it's actually the new `custom.ts` module added to `@graphiql/toolkit` that is part of this bundle. Should I create a separate issue to address this?

image

sorry @heyacherry , this test is unrelated to your PR

**Comment by B2o5T** [comment]
@acao feel free to create an additional issue about why tree shaking didn’t work for this file, I'll take a look once I come back

You can adjust the number for this PR

**Comment by acao** [comment]
@B2o5T it appears it was tree shaken after all! so `1092 modules transformed.` counts all modules, some before tree shaking

**Comment by acao** [comment]
https://github.com/graphql/graphiql/issues/3338 I've created a bug for the test, will update the test, but hopefully we can find a better way to measure this

**Comment by heyacherry** [comment]
Sorry for the deplay -- I just back to check the comments 
Great work and great thanks @acao @B2o5T 🌟🌟🌟
Let me know if need anything from me

**Comment by acao** [comment]
awesome work yourself @heyacherry !!

**Pull request merged**